staging/src/k8s.io/cloud-provider/app/webhooks.go

```diff
@@ -0,0 +1,214 @@
+			cloud:            cloud,
+		}
+	}
+	return webhookHandlers
+}
+
+func WebhookNames(webhooks map[string]WebhookConfig) []string {
+	ret := sets.StringKeySet(webhooks)
+	return ret.List()
+}
+
+func newHandler(webhooks map[string]webhookHandler) *mux.PathRecorderMux {
+	mux := mux.NewPathRecorderMux("controller-manager-webhook")
+
+	for _, handler := range webhooks {
+		mux.Handle(handler.path, handler)
+	}
+
+	return mux
+}
+
+func (h webhookHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	ctx := context.Background()
+	klog.Infof("Received validation request: %q", r.RequestURI)
+
+	start := time.Now()
+	var (
+		statusCode        int
+		err               error
+		in                *admissionv1.AdmissionReview
+		admissionResponse *admissionv1.AdmissionResponse
+	)
+	defer func() {
+		latency := time.Since(start)
+
+		if statusCode != 0 {
+			recordRequestTotal(ctx, strconv.Itoa(statusCode), h.name)
+			recordRequestLatency(ctx, strconv.Itoa(statusCode), h.name, latency.Seconds())
+			return
+		}
+
+		if err != nil {
+			recordRequestTotal(ctx, "<error>", h.name)
+			recordRequestLatency(ctx, "<error>", h.name, latency.Seconds())
+		}
+	}()
+
+	in, err = parseRequest(r)
+	if err != nil {
+		klog.Error(err)
+		statusCode = http.StatusBadRequest
+		http.Error(w, err.Error(), statusCode)
+		return
+	}
+
+	admissionResponse, err = h.admissionHandler(in.Request)
+	if err != nil {
+		e := fmt.Sprintf("error generating admission response: %v", err)
+		klog.Errorf(e)
+		statusCode = http.StatusInternalServerError
+		http.Error(w, e, statusCode)
+		return
+	} else if admissionResponse == nil {
+		e := fmt.Sprintf("admission response cannot be nil")
+		klog.Error(e)
+		statusCode = http.StatusInternalServerError
+		http.Error(w, e, statusCode)
+		return
+	}
+
+	admissionReview := &admissionv1.AdmissionReview{
+		Response: admissionResponse,
+	}
+	admissionReview.Response.UID = in.Request.UID
+	w.Header().Set("Content-Type", "application/json")
+
+	codec := codecs.EncoderForVersion(encoder, admissionv1.SchemeGroupVersion)
+	out, err := runtime.Encode(codec, admissionReview)
+	if err != nil {
+		e := fmt.Sprintf("error parsing admission response: %v", err)
+		klog.Error(e)
+		statusCode = http.StatusInternalServerError
+		http.Error(w, e, statusCode)
+		return
+	}
+
+	klog.Infof("%s", out)
+	fmt.Fprintf(w, "%s", out)
+}
+
+// parseRequest extracts an AdmissionReview from an http.Request if possible
+func parseRequest(r *http.Request) (*admissionv1.AdmissionReview, error) {
+
+	review := &admissionv1.AdmissionReview{}
+
+	if r.Header.Get("Content-Type") != "application/json" {
+		return nil, fmt.Errorf("Content-Type: %q should be %q",
+			r.Header.Get("Content-Type"), "application/json")
+	}
+
+	bodybuf := new(bytes.Buffer)
+	bodybuf.ReadFrom(r.Body)
+	body := bodybuf.Bytes()
+
+	if len(body) == 0 {
+		return nil, fmt.Errorf("admission request HTTP body is empty")
+	}
+
+	if _, _, err := deserializer.Decode(body, nil, review); err != nil {
```

Comment:
does this ensure the version sent to the webhook was actually a v1 AdmissionReview?

Reply:
staging/src/k8s.io/pod-security-admission/cmd/webhook/server/server.go might be a good pattern to copy with some defensive limiting of the body size we read in, and more efficient writing of the response

Reply:
> does this ensure the version sent to the webhook was actually a v1 AdmissionReview?

Addressed this comment

> staging/src/k8s.io/pod-security-admission/cmd/webhook/server/server.go might be a good pattern to copy with some defensive limiting of the body size we read in, and more efficient writing of the response

Can this be done in a follow up, in order to be sure we hit code freeze?

Reply:
Spoke offline about this, and we're going to do this in a followup